sst/test/src/tg2_cache_access.cc

```diff
@@ -0,0 +1,301 @@
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+
+    # Execute last loop
+    ldr x1, [x0]
+    )",
+          "2048") {
+  Parser p = Parser(capturedStdout);
+  std::vector<ParsedMemRead*> reads = p.getParsedMemReads();
+  // skip first parsed request as that one will be caused by heap address
+  // retrieval into x0.
+  uint64_t firstLoadCycleLatency = reads[1]->endCycle_ - reads[1]->startCycle_;
+  uint64_t firstLoadData = reads[1]->data_;
+  // check the last entry of the vector to avoid load executed by branch
+  // prediction.
+  uint64_t lastLoadCycleLatency =
+      reads[reads.size() - 1]->endCycle_ - reads[reads.size() - 1]->startCycle_;
+  uint64_t lastLoadData = reads[reads.size() - 1]->data_;
+
+  EXPECT_LT(lastLoadCycleLatency, firstLoadCycleLatency);
+  EXPECT_EQ(firstLoadData, lastLoadData);
+}
+
+TEST_CASE(TG2, "load_after_store_on_same_address_should_be_return_from_cache",
+          "src",
+          R"(
+    # Get heap address
+    mov x0, 0
+    mov x8, 214
+    svc #0
+
+    # Setup
+    mov x2, #2048
+
+    # Load from heap - done to determine max clock cycle without any caching.
+    ldr x1, [x0]
+
+    # Store value in x3 at address in x2
+    str x1, [x2]
+
+    # Multiple instructions to wait for store to execute.
+    # Can't be replaced by loop to ensure that caching was
+    # established by store
+    mov x4, #0
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+
+    # last load
+    ldr x4, [x2]
+
+    )",
+          "1024") {
+  Parser p = Parser(capturedStdout);
+  std::vector<ParsedMemRead*> reads = p.getParsedMemReads();
+  // skip first parsed request as that one will be caused by heap address
+  // retrieval into x0.
+  uint64_t firstLoadCycleLatency = reads[1]->endCycle_ - reads[1]->startCycle_;
+  uint64_t firstLoadData = reads[1]->data_;
+  uint64_t lastLoadCycleLatency = reads[2]->endCycle_ - reads[2]->startCycle_;
+  uint64_t lastLoadData = reads[2]->data_;
+
+  EXPECT_LT(lastLoadCycleLatency, firstLoadCycleLatency);
+  EXPECT_EQ(1024, firstLoadData);
+  EXPECT_EQ(1024, lastLoadData);
+}
```

Comment:
Do these two tests not do essentially the same thing? i.e. couldn't the first test's functionality be verified by the second test?

Reply:
In theory, yes but there is a small difference: The first test checks if load requests are cached i.e if something was loaded from the initialised process image, without having been stored first. The second test checks if the value being stored is cached simultaneously (write-through)
The load address in the second test is the same as the store address, whereas in the first test the address for all load requests is the heap address.

Reply:
I see, thanks for explaining

Reply:
There is a small typo in the first test case, the comment for the first sequence of instructions is misleading. Will change it